test/test_nn.py

```diff
@@ -35,6 +35,28 @@
 if TEST_SCIPY:
     from scipy import stats
 
+ALL_TENSORTYPES = [torch.FloatTensor,
+                   torch.DoubleTensor,
+                   torch.HalfTensor]
+
+NO_HALF_TENSORTYPES = [torch.FloatTensor,
+                       torch.DoubleTensor]
+
+DOUBLE_TENSORTYPES = [torch.DoubleTensor]
+
+type2prec = {'FloatTensor': 1e-5,
+             'DoubleTensor': 1e-5,
+             'HalfTensor': 1e-2}
+
+# Used for run the same test with different tensor types
+def repeat_test_for_types(dtypes):
+    def repeatHelper(f):
+        def callHelper(*args):
+            for dtype in dtypes:
```

Comment:
One suggestion would be to use `subTest` so that we see for which type the failure occured (but you need to guard it so that it's only used in Python 3.4+